### DISCUSSION TITLE
Confirming which classes are using FieldBackedProvider

### DISCUSSION THREAD

**Discussion by anuraaga** [open-telemetry/opentelemetry-java-instrumentation#2612] [Q&A]
Do we have a way to confirm (ideally integration test) which classes are correctly being instrumented with a field by FieldBackedProvider? It's mostly in the context of #2601, by using more stuff in the bootstrap classloader I think there's always a chance of accidentally classloading something too early. How do we confirm we don't introduce such a regression?

**Comment by trask** [comment]
hey, I was just thinking about this yesterday 😁

I think we could add a test flag that triggers a failure if we aren't using the good field-based implementation (using the fallback weak map), and then if we find there are a few classes, we can allow only those specific classes to use the fallback weak map under the test flag.

I'm not sure if this would have been possible before #1643, because the tests relied on retransforming lots of classes, but I think it should be possible now.
